### PR TITLE
metric logger - grpo metrics (3/N)

### PR DESCRIPTION
**Pull request by felipemello1**
Adds metrics to GRPO. Also adds conftest.py to ignore record_metrics for unit testing.

```
WandbBackend: Logged 86 metrics at step 1
=== [global_logger_fF8h_r0] - METRICS STEP 1 ===
  buffer/add/count_episodes_added: 24.0
  buffer/evict/avg_policy_staleness: 0.0
  buffer/evict/max_policy_staleness: 0.0
  buffer/sample/avg_buffer_utilization: 16.0
  buffer/sample/avg_buffer_utilization_pct: 100.0
  buffer/sample/count_sample_requests: 239.0
  buffer_perf/sample/duration_avg_s: 5.8068605490558815e-05
  buffer_perf/sample/duration_max_s: 0.004135082941502333
  dataset/sample/avg_sample_len: 525.25
  dataset/sample/count_samples_generated: 4.0
  main/rollout/avg_advantage: -7.450580596923828e-09
  main/rollout/count_rollout_iterations: 3.0
  main/rollout/std_advantage: 0.9332931327148772
  main/training/count_training_iterations: 1.0
  main_perf/rollout/advantage_computation/duration_avg_s: 0.05832476494833827
  main_perf/rollout/advantage_computation/duration_max_s: 0.06426444882526994
  main_perf/rollout/compute_logprobs/duration_avg_s: 0.20812372273455063
  main_perf/rollout/compute_logprobs/duration_max_s: 0.24763409001752734
  main_perf/rollout/data_loading/duration_avg_s: 0.145997637261947
  main_perf/rollout/data_loading/duration_max_s: 0.43461281480267644
  main_perf/rollout/policy_generation/duration_avg_s: 3.4080680563735464
  main_perf/rollout/policy_generation/duration_max_s: 4.955212695058435
  main_perf/rollout/reference_model_forward/duration_avg_s: 11.607943433026472
  main_perf/rollout/reference_model_forward/duration_max_s: 12.128578653093427
  main_perf/rollout/reward_evaluation/duration_avg_s: 0.010726675235976776
  main_perf/rollout/reward_evaluation/duration_max_s: 0.017631276044994593
  main_perf/rollout/total_duration_avg_s: 15.439195216012498
  main_perf/rollout/total_duration_max_s: 17.847949991934
  policy/generate/avg_tokens_generated: 508.90625
  policy/generate/count_generate_requests: 4.0
  policy/generate/count_requests: 4.0
  policy/generate/count_sequences_completed: 32.0
  policy/generate/sum_tokens_generated: 16285.0
  policy/update_weights/count_weight_updates: 1.0
  policy_perf/generate/generate/duration_avg_s: 4.422806396484376
  policy_perf/generate/generate/duration_max_s: 9.698044921875
  policy_perf/generate/process_inputs/duration_avg_s: 0.0059196240603923795
  policy_perf/generate/process_inputs/duration_max_s: 0.01985408020019531
  policy_perf/generate/prompt_truncation/duration_avg_s: 3.808000119170174e-05
  policy_perf/generate/prompt_truncation/duration_max_s: 0.00014249600470066072
  policy_perf/generate/total_duration_avg_s: 4.434622764600382
  policy_perf/generate/total_duration_max_s: 9.699662105880911
  reference_perf/forward/avg_sequence_length: 1024.0
  reference_perf/forward/count_forward_passes: 4.0
  reference_perf/forward/forward/duration_avg_s: 0.1228512477874756
  reference_perf/forward/forward/duration_max_s: 0.19166079711914064
  reference_perf/forward/garbage_collection/duration_avg_s: 0.00021811200631782413
  reference_perf/forward/garbage_collection/duration_max_s: 0.0005756480097770691
  reference_perf/forward/memory_delta_peak_start_avg_gb: 3.88580322265625
  reference_perf/forward/memory_peak_max_gb: 15.53714370727539
  reference_perf/forward/to_device/duration_avg_s: 0.00043163999915122984
  reference_perf/forward/to_device/duration_max_s: 0.00048412799835205076
  reference_perf/forward/total_duration_avg_s: 0.12350442379293963
  reference_perf/forward/total_duration_max_s: 0.1921454051337205
  reward/evaluate_response/avg_MathReward_reward: 0.134375
  reward/evaluate_response/avg_ThinkingReward_reward: 0.4749999999999997
  reward/evaluate_response/avg_total_reward: 0.30468749999999994
  reward/evaluate_response/count_MathReward_calls: 32.0
  reward/evaluate_response/count_ThinkingReward_calls: 32.0
  reward/evaluate_response/std_MathReward_reward: 0.22722424909106842
  reward/evaluate_response/std_ThinkingReward_reward: 0.3799671038392665
  reward/evaluate_response/sum_MathReward_reward: 4.3
  reward/evaluate_response/sum_ThinkingReward_reward: 15.19999999999999
  rl_trainer_perf/push_weights/dcp_save/duration_avg_s: 12.9625517578125
  rl_trainer_perf/push_weights/dcp_save/duration_max_s: 12.9625517578125
  rl_trainer_perf/push_weights/flatten_state_dict/duration_avg_s: 0.0003195199966430664
  rl_trainer_perf/push_weights/flatten_state_dict/duration_max_s: 0.0003195199966430664
  rl_trainer_perf/push_weights/memory_delta_peak_start_avg_gb: 2.86102294921875e-06
  rl_trainer_perf/push_weights/memory_peak_max_gb: 22.771067142486572
  rl_trainer_perf/push_weights/to_hf/duration_avg_s: 0.0008420799970626831
  rl_trainer_perf/push_weights/to_hf/duration_max_s: 0.0008420799970626831
  rl_trainer_perf/push_weights/total_duration_avg_s: 12.96401162981987
  rl_trainer_perf/push_weights/total_duration_max_s: 12.96401162981987
  trainer/avg_grpo_loss: 1.4528632164001465e-06
  trainer/count_training_steps: 1.0
  trainer/learning_rate: 0.001
  trainer_perf/memory_delta_peak_start_avg_gb: 30.207598209381104
  trainer_perf/memory_peak_max_gb: 37.778372287750244
  trainer_perf/step/forward_backward/duration_avg_s: 0.923697265625
  trainer_perf/step/forward_backward/duration_max_s: 0.923697265625
  trainer_perf/step/optimizer_step/duration_avg_s: 0.041217311859130856
  trainer_perf/step/optimizer_step/duration_max_s: 0.041217311859130856
  trainer_perf/step/save_checkpoint/duration_avg_s: 0.001040768027305603
  trainer_perf/step/save_checkpoint/duration_max_s: 0.001040768027305603
  trainer_perf/step/total_duration_avg_s: 0.9662981294989585
  trainer_perf/step/total_duration_max_s: 0.9662981294989585
==============================
```

### CODE REVIEW

File: apps/grpo/main.py [hunk 1]
```diff
@@ -315,45 +417,75 @@ async def continuous_rollouts():
...
+                record_metric(
+                    "main/rollout/std_advantage", advantage, ReductionType.STD
+                )
+            timer.step("advantage_computation")
 
             rollout_count += 1
+            record_metric("main/rollout/count_rollout_iterations", 1, ReductionType.SUM)
+            timer.end()
 
     async def continuous_training():
```
Comment (felipemello1) [inline apps/grpo/main.py:446]: @Jack-Khuu

File: .gitignore [hunk 2]
```diff
@@ -196,3 +196,4 @@ assets/wheels/vllm*.whl
 
 # DCP artifacts
 model_state_dict/
+forge_dcp_tmp/
```
Comment (allenwang28) [inline .gitignore:201]: where did this come from? If the name has shifted to model_state_dict we should just remove the above line

File: src/forge/actors/reference_model.py [hunk 3]
```diff
@@ -88,9 +90,31 @@ async def setup(self):
         self.model.eval()
 
     @endpoint
+    @trace(
+        "reference_perf/forward",
+        track_time=False,
+        track_memory=True,
+        time_with_gpu=True,
```
Comment (allenwang28) [inline src/forge/actors/reference_model.py]: ok another follow up I should've caught earlier - it's confusing that we're saying `track_time=False` because we want `time_with-gpu=True` lol, maybe we should change to
```
@trace(
    "reference_perf/forward",
    track_memory=True,
    timer="cuda",
)
```

File: apps/grpo/main.py [hunk 4]
```diff
@@ -324,48 +390,61 @@ async def continuous_rollouts():
...
                 episode.ref_logprobs = ref_logprobs[i]
             del ref_logits, ref_logprobs, input_ids
+            t.step("compute_logprobs")
 
             # Calculate advantages and add to replay buffer
             advantages = await compute_advantages.compute.call_one(group)
             for episode, advantage in zip(group.episodes, advantages):
                 episode.advantage = advantage
-                await replay_buffer.add.call_one(episode)
+                await replay_buffer.add.route(episode)
```
Comment (allenwang28) [inline apps/grpo/main.py]: i think this overrides Danning's recent change, unfortunate merge conflict. Can you "revert" to the actor semantics?

File: apps/grpo/main.py [hunk 5]
```diff
@@ -379,6 +458,8 @@ async def continuous_training():
         training_task.cancel()
     finally:
         print("Shutting down...")
+        await asyncio.gather(mlogger.shutdown.call_one())
```
Comment (allenwang28) [inline apps/grpo/main.py]: can we just add this into the asyncio.gather group below?

File: src/forge/actors/policy.py [hunk 6]
```diff
@@ -190,6 +192,8 @@ async def launch(  # pyright: ignore[reportIncompatibleMethodOverride]
             engine_config = EngineConfig.from_dict(engine_config)
 
         vllm_config = engine_config.create_vllm_config()
+        # TODO (felipemello): LocalFetcherActor doesnt spawn with this, so cannot
+        # do logging within PolicyWorker
```
Comment (allenwang28) [inline src/forge/actors/policy.py:196]: hmm interesting, i would expect this to happen since it's also going through get_proc_mesh...

File: apps/grpo/main.py [hunk 7]
```diff
@@ -324,48 +390,61 @@ async def continuous_rollouts():
...
+            batch = await replay_buffer.sample.route(curr_policy_version=training_step)
             if batch is None:
+                # Track time waiting for buffer
                 await asyncio.sleep(0.1)
             else:
+                t.step("waiting_for_buffer")
+
                 inputs, targets = batch
-                loss = await trainer.train_step.call_one(inputs, targets)
+                loss = await trainer.train_step.route(inputs, targets)
```
Comment (felipemello1) [inline apps/grpo/main.py]: double checking these changes. They shouldnt have happened.

File: src/forge/actors/trainer.py [hunk 8]
```diff
@@ -319,30 +353,32 @@ async def _push_weights_DEPRECATED(  # noqa: N802
...
-        logger.info(
-            f"Completed weights push to {key} in {end_time - start_time:.2f} seconds "
-            f"(hg to vllm conversion: {conversion_time - start_time:.2f}s, tranport time: {end_time - conversion_time:.2f})"
-        )
 
     @endpoint
+    @trace(prefix="rl_trainer_perf/push_weights", track_time=False, track_memory=True)
     async def push_weights(self, policy_version: int) -> None:
         """Push weights to torchstore in HF format."""
+        t = Tracer("rl_trainer_perf/push_weights", time_with_gpu=True)
```
Comment (allenwang28) [inline src/forge/actors/trainer.py]: why do we need both the decorator and a tracer object?

Comment (felipemello1) [reply src/forge/actors/trainer.py]: we dont anymore. Before Tracer was just Timer, no memory tracking. Let me update it.

File: apps/grpo/main.py (hunk 4 above)
Comment (felipemello1) [reply apps/grpo/main.py]: yes, looking into it now. Just realized it happened.

File: src/forge/actors/reference_model.py (hunk 3 above)
Comment (felipemello1) [reply src/forge/actors/reference_model.py]: agreed. Before i had "use_cuda", but i wanted it to be gpu agnostic, i.e. AMD, TPU, etc.

How about:
```
@trace(
    "reference_perf/forward",
    track_memory=True,
    timer=Enum['gpu', 'cpu', None]
)
```

Comment (allenwang28) [reply src/forge/actors/reference_model.py]: i think just string for cpu/gpu is fine for now

File: .gitignore (hunk 2 above)
Comment (felipemello1) [reply .gitignore:201]: i think some older code (used in other scripts) still uses it. Leaving it for now

File: apps/grpo/main.py (hunk 5 above)
Comment (felipemello1) [reply apps/grpo/main.py]: no, i actually added an asyncio.sleep to logger has time to shutdown, otherwise wandb processes stay alive :/

Comment (felipemello1) [reply apps/grpo/main.py]: will add to the follow up list

File: apps/grpo/main.py [hunk 9]
```diff
@@ -171,9 +173,47 @@ async def evaluate_response(self, prompt: str, response: str, target: str) -> fl
         for reward_fn in self.reward_functions:
             reward = reward_fn(prompt, response, target)
             total_rewards += reward
-        return total_rewards / len(self.reward_functions)
+
+            # Get a name for the reward function (works for classes, functions, lambdas)
+            reward_fn_name = getattr(
+                reward_fn, "__name__", reward_fn.__class__.__name__
```
Comment (allenwang28) [inline apps/grpo/main.py:184]: @joecummings should we do this or set a name explicitly?

File: src/forge/actors/policy.py (hunk 6 above)
Comment (allenwang28) [reply src/forge/actors/policy.py:196]: is this why you didn't add metrics or tracing to the workers?

File: src/forge/actors/reference_model.py [hunk 10]
```diff
@@ -88,9 +90,30 @@ async def setup(self):
...
+
+        # Record reference model metrics
+        record_metric("reference_perf/forward/count_forward_passes", 1, Reduce.SUM)
+        record_metric(
+            "reference_perf/forward/avg_sequence_length",
+            input_ids.shape[1],
+            Reduce.MEAN,
+        )
+
+        t = Tracer("reference_perf/forward", timer="gpu")
```
Comment (allenwang28) [inline src/forge/actors/reference_model.py]: if we're creating a tracer anyways, I'd prefer not to have the function decorated

File: src/forge/actors/replay_buffer.py [hunk 11]
```diff
@@ -91,9 +115,9 @@ async def evict(self, curr_policy_version: int) -> None:
         Args:
             curr_policy_version (int): The current policy version.
         """
-        self._evict(curr_policy_version)
+        await self._evict(curr_policy_version)
 
-    def _evict(self, curr_policy_version: int) -> None:
+    async def _evict(self, curr_policy_version: int) -> None:
```
Comment (allenwang28) [inline src/forge/actors/replay_buffer.py]: why did this need to become async?

Comment (felipemello1) [reply src/forge/actors/replay_buffer.py]: Will remove. Maybe because before i was doing `await record_metrics`?

File: src/forge/actors/policy.py (hunk 6 above)
Comment (felipemello1) [reply src/forge/actors/policy.py:196]: yes, need to investigate it. I could add where it gets called. Not sure if i did

Comment (allenwang28) [reply src/forge/actors/policy.py:196]: ok got it
